relay-quotas/src/rate_limit.rs

```diff
@@ -413,25 +415,52 @@ impl CachedRateLimits {
     /// Adds a limit to this collection.
     ///
     /// See also: [`RateLimits::add`].
-    pub fn add(&mut self, limit: RateLimit) {
-        self.0.add(limit);
+    pub fn add(&self, limit: RateLimit) {
+        self.0.rcu(|current| {
+            let mut current = current.as_ref().clone();
+            current.add(limit.clone());
+            current
+        });
     }
 
     /// Merges more rate limits into this instance.
     ///
     /// See also: [`RateLimits::merge`].
-    pub fn merge(&mut self, rate_limits: RateLimits) {
-        for limit in rate_limits {
-            self.add(limit)
-        }
+    pub fn merge(&self, limits: RateLimits) {
+        self.0.rcu(|current| {
+            let mut current = current.as_ref().clone();
+            for limit in limits.clone() {
+                current.add(limit)
+            }
+            current
+        });
     }
 
     /// Returns a reference to the contained rate limits.
     ///
     /// This call gurantuess that at the time of call no returned rate limit is expired.
-    pub fn current_limits(&mut self) -> &RateLimits {
-        self.0.clean_expired();
-        &self.0
+    pub fn current_limits(&self) -> Arc<RateLimits> {
+        let now = Instant::now();
+
+        let mut current = self.0.load();
+        while current.iter().any(|rl| rl.retry_after.expired_at(now)) {
+            let new = {
+                let mut new = current.as_ref().clone();
+                new.clean_expired();
+                Arc::new(new)
+            };
+
+            let prev = self.0.compare_and_swap(&*current, Arc::clone(&new));
+
+            // If there was a swap, we know `new` is now stored and the most recent value.
+            if Arc::ptr_eq(&current, &prev) {
+                return new;
+            }
+
+            current = prev;
+        }
+
+        arc_swap::Guard::into_inner(current)
```

Comment:
I think we can also get a way with a very simple `Mutex` variant (like `ReservoirCounters`), I chose `ArcSwap` because I just had it at hand and liked the idea of being fully lock-free.

Reply:
Missed your second part:

> Do we know anything about the worst case behavior of the loop? Should we log an error and bail if it runs for more than a few iterations? I know rcu uses a similar loop, but it still makes me nervous.
It must make progress at some point and ideally it never runs more than once. Either the swap succeeds, or it does not but then somebody else already cleaned the expired limits and we can proceed with those.